paddle/operators/lod_reset_op.cc

```diff
@@ -0,0 +1,101 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#include "paddle/operators/lod_reset_op.h"
+
+namespace paddle {
+namespace operators {
+
+class LoDResetOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    // input check
+    PADDLE_ENFORCE(ctx->HasInput("X"),
+                   "Input(X) of LoDResetOp should not be null.");
+    PADDLE_ENFORCE(ctx->HasOutput("Out"),
+                   "Output(Out) of LoDResetOp should not be null.");
+
+    auto level0 = ctx->Attrs().Get<std::vector<int>>("target_lod_0");
+    PADDLE_ENFORCE(level0.size() > 1UL,
+                   "The size of Attr(target_lod_0) should be greater than 1.");
+    auto x_dims = ctx->GetInputDim("X");
+    PADDLE_ENFORCE(level0[0] == 0,
+                   "Attr(target_lod_0) should be a vector starting from 0.");
+    PADDLE_ENFORCE(level0.back() == x_dims[0],
+                   "Attr(target_lod_0) should be a vector end with the "
+                   "first dimension of Input(X).");
+    for (size_t i = 0; i < level0.size() - 1; ++i) {
+      PADDLE_ENFORCE(level0[i + 1] > level0[i],
+                     "Attr(target_lod_0) should be an ascending vector.");
+    }
+    ctx->SetOutputDim("Out", x_dims);
+  }
+};
+
+class LoDResetOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  LoDResetOpMaker(framework::OpProto *proto,
+                  framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "The input tensor of lod_reset operator.");
+    AddOutput("Out", "The output tensor of lod_reset operator.");
+    AddAttr<std::vector<int>>("target_lod_0",
+                              "Target level 0 LoD of "
+                              "lod_reset operator.");
```

Comment:
Now, the attributes are usually fixed during training, that is to say, it was passed to the network when creating the graph. I'm not sure whether the attributes can be changed during training in the future.  And since the batch size and the sequence length may be varying during training,  I think it's better to pass LoD as an input, not attributes. Sine the input can be calculated by others operator.

Reply:
Done